code/datums/beam.dm

```diff
@@ -0,0 +1,190 @@
+/** # Beam Datum and Effect
+ * **IF YOU ARE LAZY AND DO NOT WANT TO READ, GO TO THE BOTTOM OF THE FILE AND USE THAT PROC!**
+ *
+ * This is the beam datum! It's a really neat effect for the game in drawing a line from one atom to another.
+ * It has two parts:
+ * The datum itself which manages redrawing the beam to constantly keep it pointing from the origin to the target.
+ * The effect which is what the beams are made out of. They're placed in a line from the origin to target, rotated towards the target and snipped off at the end.
+ * These effects are kept in a list and constantly created and destroyed (hence the proc names draw and reset, reset destroying all effects and draw creating more.)
+ *
+ * You can add more special effects to the beam itself by changing what the drawn beam effects do. For example you can make a vine that pricks people by making the beam_type
+ * include a crossed proc that damages the crosser. Examples in venus_human_trap.dm
+*/
+/datum/beam
+	///where the beam goes from
+	var/atom/origin = null
+	///where the beam goes to
+	var/atom/target = null
+	///list of beam objects. These have their visuals set by the visuals var which is created on starting
+	var/list/elements = list()
+	///icon used by the beam.
+	var/icon
+	///icon state of the main segments of the beam
+	var/icon_state = ""
+	///The beam will qdel if it's longer than this many tiles.
+	var/max_distance = 0
+	///the objects placed in the elements list
+	var/beam_type = /obj/effect/ebeam
+	///This is used as the visual_contents of beams, so you can apply one effect to this and the whole beam will look like that. never gets deleted on redrawing.
+	var/obj/effect/ebeam/visuals
+	///will the origin object always turn to face the target?
+	var/always_turn = FALSE
+
+/datum/beam/New(beam_origin,beam_target,beam_icon='icons/effects/beam.dmi',beam_icon_state="b_beam",time=BEAM_INFINITE_DURATION,maxdistance=INFINITY,btype = /obj/effect/ebeam)
+	origin = beam_origin
```

Comment:
These assignments should be using src.var rather than having two different names for the same variables.